### PR TITLE
Smartpooling

### PR DESCRIPTION
**Pull request by jdzikowski**
Initial implementation of smartpooling operation.

### CODE REVIEW

File: fairseq/models/wav2vec/wav2vec2_scribblelens.py [hunk 1]
```diff
@@ -403,6 +417,11 @@ def upgrade_state_dict_named(self, state_dict, name):
         """Upgrade a (possibly old) state dict for new versions of fairseq."""
         return state_dict
 
+    def to(self, *args, **kwargs):
+        self = super().to(*args, **kwargs) 
+        self.smartpooling_filters = self.smartpooling_filters.to(*args, **kwargs)
```
Comment (janchorowski) [inline fairseq/models/wav2vec/wav2vec2_scribblelens.py]: To nie będzie potrzebne, jesli smartpooling filters beda zadeklarowane jako buffer, (tensor, ale nieuczony) albo parameter (tensor, uczony przez backprop)

File: fairseq/models/wav2vec/wav2vec2_scribblelens.py [hunk 2]
```diff
@@ -525,6 +544,43 @@ def compute_preds(self, x, y, negatives):
 
         return logits
 
+    def smartpool(self, features, padding_mask=None):
+        features_tmp = F.pad(features,(0,0,1,0))
+        new_lens = (features_tmp[:,1:,:] - features_tmp[:,:-1,:]).abs().sum(dim=2)
+        new_lens = new_lens / new_lens.sum(1, keepdim=True) * (features.size(1) / self.smartpooling_factor) # Reducing the original length T by some factor
+        
+        features, interp_weights = self.warp(features, new_lens)
+        if padding_mask is not None :
```
Comment (janchorowski) [inline fairseq/models/wav2vec/wav2vec2_scribblelens.py]: Padding mask oznacza które kolumny są poza danymi, więc paddingu nie skracamy, to co trzeba zrobić to skrócić te cześci unpadded, a potem sprawdzić który jest najdłuższy i spadować resztę.

**Comment by janchorowski** [comment]
Tak na szybko, to to chyba jest OK, natomiast muszę do tego poolingu zrobić jakąś wizualizację żeby zobaczyć jak on działa. To co na pewno chcemy (nie wiem czy tak jest), to żeby pochodna propagowała się do wegtora z nowymi długościami - wyobrażam sobie, że to działa tak, że jak np mamy nowe długości:
1,1,1
i skracamy do dwóch elementów, to 
wiemy że nowy element 0 wziął się ze starych 0 i 1 i że jeśli chcilibyśmy tam np. zwiększyć udział starego 0, to musimy zminić wektor długości na 1+eps, 1, 1.
Bez tego, to może się nie uczyć
